### PR TITLE
Introduce data_dictionary module

### PR DESCRIPTION
**Pull request by avikivity**
The full user-defined structure of the database (keyspaces,
tables, user-defined types, and similar metadata, often known
as the schema in other databases) is needed by much of the
front-end code. But in Scylla it is deeply intertwined with the
replica data management code - ::database, ::keyspace, and
::table. Not only does the front-end not need data access, it
cannot get correct data via these objects since they represent
just one replica out of many.

This dual-role is a frequent cause of recompilations. It was solved
to some degree by forward declarations, but there is still a lot
of incidental dependencies.

To solve this, we introduce a data_dictionary module (and
namespace) to exclusively deal with greater schema metadata.
It is an interface, with a backing implementation by the existing code,
so it doesn't add a new source of truth. The plan is to allow mock
implementations for testing as well.

Test: unit (dev, release, debug).

### CODE REVIEW

File: data_dictionary/data_dictionary.cc
Comment (nyh) [inline data_dictionary/data_dictionary.cc:173]: Maybe you should add here a FIXME that this method should eventually be removed, as you explained in the commit message? A couple of years from now nobody will remember that the intention was to remove this method, and use it in more places :-(

File: auth/common.cc
Comment (nyh) [inline auth/common.cc:67]: The change in this source file is a bit underwhelming... Wasn't the intention that we can get rid of `#include <database.hh>`? 
Or maybe this will just come later and you started with just the simplest transformations in each source file?

**Review by nyh — CHANGES_REQUESTED** [review]
Looks like a good idea and a good implementation, but I have a couple of questions:

I know that this is bike-shedding par-excellence, but I'm not a big fan of the name "data_dictionary" - I don't think there's any obvious meaning here to the word "dictionary" (after all, doesn't a dictionary contain... data?), and it's such a long word...
Can't we call the namespace just "metadata"? So instead of `data_dictionary::keyspace_metadata` we'll have `metadata::keyspace_metadata`. Or even shorter, just `metadata::keyspace` - just like you have `database` and not `database_metadata`.?
Alternatively, if you want a long name, maybe `data_definition` instead of `data_dictionary`? or `data_structure`?

Another question I have that isn't exactly clear in what you wrote in comments or commit message - is what exactly do these "metadata" objects do in respect to versioning and lifetimes. If I hold a "table" object, and the table gets changed or even deleted in the background, what happens?

**Comment by gleb-cloudius** [comment]
This will conflict with "Move schema altering statement to raft" series. I do not see anything urgent here that cannot wait for that series to be merged.

**Comment by avikivity** [comment]
> Looks like a good idea and a good implementation, but I have a couple of questions:
> 
> I know that this is bike-shedding par-excellence, but I'm not a big fan of the name "data_dictionary" - I don't think there's any obvious meaning here to the word "dictionary" (after all, doesn't a dictionary contain... data?), and it's such a long word... Can't we call the namespace just "metadata"? So instead of `data_dictionary::keyspace_metadata` we'll have `metadata::keyspace_metadata`. Or even shorter, just `metadata::keyspace` - just like you have `database` and not `database_metadata`.? Alternatively, if you want a long name, maybe `data_definition` instead of `data_dictionary`? or `data_structure`?

It's the standard name for this sort of thing: https://en.wikipedia.org/wiki/Data_dictionary.

> 
> Another question I have that isn't exactly clear in what you wrote in comments or commit message - is what exactly do these "metadata" objects do in respect to versioning and lifetimes. If I hold a "table" object, and the table gets changed or even deleted in the background, what happens?

A data_dictionary::table object is similar to a C++ reference. If the table is changed, this is reflected in the data_dictionary::table. If the table is deleted, you end up holding a dangling reference.

A data_dictionary::database object is valid for the whole duration of the program (at least until the ::database object is destroyed).

Callers should manage it by not holding references across continuations, or arranging for invalidation notifications.

**Comment by avikivity** [comment]
> This will conflict with "Move schema altering statement to raft" series. I do not see anything urgent here that cannot wait for that series to be merged.

I don't mind rebasing this series after yours is merged.

File: auth/common.cc
Comment (avikivity) [reply auth/common.cc:67]: I focused on cql3/. I'll tackle other coordinator-side modules later (thrift, transport, redis)

File: data_dictionary/data_dictionary.cc
Comment (avikivity) [reply data_dictionary/data_dictionary.cc:173]: There is a comment in the declaration.

**Comment by nyh** [comment]
> > Looks like a good idea and a good implementation, but I have a couple of questions:
> > I know that this is bike-shedding par-excellence, but I'm not a big fan of the name "data_dictionary" - I don't think there's any obvious meaning here to the word "dictionary" (after all, doesn't a dictionary contain... data?), and it's such a long word... Can't we call the namespace just "metadata"? So instead of `data_dictionary::keyspace_metadata` we'll have `metadata::keyspace_metadata`. Or even shorter, just `metadata::keyspace` - just like you have `database` and not `database_metadata`.? Alternatively, if you want a long name, maybe `data_definition` instead of `data_dictionary`? or `data_structure`?
> 
> It's the standard name for this sort of thing: https://en.wikipedia.org/wiki/Data_dictionary.

I'm not a database expert, so perhaps unsurprisingly (?) I never heard this term used. And as someone who never heard this term, it didn't make any sense to me. This "Data Dictionary" does not contain "data" (it contains metadata), and it's not clear what "dictionary" refers to. I do see now that Oracle uses this name: https://docs.oracle.com/cd/B10501_01/server.920/a96524/c05dicti.htm
But why don't we just call it metadata? Especially when two of the classes are already called keyspace_metadata and table_metadata?
https://dba.stackexchange.com/questions/131293/alternate-name-of-data-dictionary also suggests the name "catalog".

> 
> > Another question I have that isn't exactly clear in what you wrote in comments or commit message - is what exactly do these "metadata" objects do in respect to versioning and lifetimes. If I hold a "table" object, and the table gets changed or even deleted in the background, what happens?
> 
> A data_dictionary::table object is similar to a C++ reference. If the table is changed, this is reflected in the data_dictionary::table. If the table is deleted, you end up holding a dangling reference.

Ok, I think this is not obvious. For example, our "schema_ptr" is different - it doesn't change and it doesn't "dangle".

> 
> A data_dictionary::database object is valid for the whole duration of the program (at least until the ::database object is destroyed).
> 
> Callers should manage it by not holding references across continuations, or arranging for invalidation notifications.

**Comment by avikivity** [comment]
> > > Looks like a good idea and a good implementation, but I have a couple of questions:
> > > I know that this is bike-shedding par-excellence, but I'm not a big fan of the name "data_dictionary" - I don't think there's any obvious meaning here to the word "dictionary" (after all, doesn't a dictionary contain... data?), and it's such a long word... Can't we call the namespace just "metadata"? So instead of `data_dictionary::keyspace_metadata` we'll have `metadata::keyspace_metadata`. Or even shorter, just `metadata::keyspace` - just like you have `database` and not `database_metadata`.? Alternatively, if you want a long name, maybe `data_definition` instead of `data_dictionary`? or `data_structure`?

I prefer reusing accepted terms (and pointing people to existing documentation) than inventing new names and then having to document and explain them.

"metadata" is a generic term. If you tell someone that X is metadata, it doesn't tell them what type of metadata it contains.

In comparison, data dictionary is specific, though it may be surprising to people who are unfamiliar.



> > 
> > 
> > It's the standard name for this sort of thing: https://en.wikipedia.org/wiki/Data_dictionary.
> 
> I'm not a database expert, so perhaps unsurprisingly (?) I never heard this term used. And as someone who never heard this term, it didn't make any sense to me. This "Data Dictionary" does not contain "data" (it contains metadata), and it's not clear what "dictionary" refers to. I do see now that Oracle uses this name: https://docs.oracle.com/cd/B10501_01/server.920/a96524/c05dicti.htm But why don't we just call it metadata? Especially when two of the classes are already called keyspace_metadata and table_metadata? https://dba.stackexchange.com/questions/131293/alternate-name-of-data-dictionary also suggests the name "catalog".



> 
> > > Another question I have that isn't exactly clear in what you wrote in comments or commit message - is what exactly do these "metadata" objects do in respect to versioning and lifetimes. If I hold a "table" object, and the table gets changed or even deleted in the background, what happens?
> > 
> > 
> > A data_dictionary::table object is similar to a C++ reference. If the table is changed, this is reflected in the data_dictionary::table. If the table is deleted, you end up holding a dangling reference.
> 
> Ok, I think this is not obvious. For example, our "schema_ptr" is different - it doesn't change and it doesn't "dangle".
> 
> > A data_dictionary::database object is valid for the whole duration of the program (at least until the ::database object is destroyed).
> > Callers should manage it by not holding references across continuations, or arranging for invalidation notifications.

**Comment by nyh** [comment]
> I prefer reusing accepted terms (and pointing people to existing documentation) than inventing new names and then having to document and explain them.
> 
> "metadata" is a generic term. If you tell someone that X is metadata, it doesn't tell them what type of metadata it contains.

Exactly - `metadata::keyspace` is the metadata about the keyspace, `metadata::table` is the metadata about the table, and so on. The `metadata` here is just a namespace which says it's about metadata.

As I noted, we already used the name "metadata", as in "user_types_metdata" and "keyspace_metadata". My suggestion is just to move this word "metadata" which we already used, to a keyspace name.

> In comparison, data dictionary is specific, though it may be surprising to people who are unfamiliar.

Even if someone is familiar with, e.g., its Oracle implementation, doesn't it imply too many connotations? Like that looks like data? (it seems Oracle's implementation is like Scylla's "system tables"?)

**Comment by avikivity** [comment]
Yes, metadata::table is metadata about the table. But which metadata? You have to guess. There can be tons of metadata about a table, like data distribution statistics. This isn't part of the data dictionary.

"data dictionary" is not an Oracle specific term, it's used across the industry.

**Comment by kostja** [comment]
* Gleb Natapov ***@***.***> [21/12/12 17:10]:
> This will conflict with "Move schema altering statement to raft" series. I do not see anything urgent here that cannot wait for that series to be merged.
> 

afics it's already merged.


-- 
Konstantin Osipov, Moscow, Russia

**Comment by gleb-cloudius** [comment]
On Sun, Dec 12, 2021 at 12:19:02PM -0800, Konstantin Osipov wrote:
> * Gleb Natapov ***@***.***> [21/12/12 17:10]:
> > This will conflict with "Move schema altering statement to raft" series. I do not see anything urgent here that cannot wait for that series to be merged.
> > 
> 
> afics it's already merged.
> 
Nope.

--
			Gleb.

File: cql3/statements/create_keyspace_statement.cc
Comment (psarna) [inline cql3/statements/create_keyspace_statement.cc:212]: I can't place a comment on the commit message, so I'll do so here. Typo: s/moce/move

File: data_dictionary/data_dictionary.cc
Comment (psarna) [inline data_dictionary/data_dictionary.cc:22]: commit message typo: s/onlu/only

Comment (psarna) [reply data_dictionary/data_dictionary.cc:22]: commit message typo 2: s/will be addressed layer/will be addressed later/

File: database.cc
Comment (psarna) [inline database.cc:72]: commit message typo: s/gnarly bits the compose/gnarly bits that compose/

**Review by psarna — COMMENTED** [review]
Looks good so far, I only found a few typos in the commit messages; I'm yet to review the largest commit though

**Comment by avikivity** [comment]
v2:
 - rebased and retested
 - corrected numerous spelling errors in commit messages
 - kept the data_dictionary terminology, I think it's better to use existing specific terms than more general ones

**Review by nyh — APPROVED** [review]

**Pull request closed without merging**
